docs/designs/keras_callback.md

```diff
@@ -0,0 +1,301 @@
+# Design to Support Custom Callback Using Keras API
+
+This document describes the design for supporting callback to customize the behavior of model during training, evaluation and inference in ElasticDL.
+
+## Motivation
+
+In deep learning, we generally need to customize the behavior of model during training, evaluation or
+inference, including reading/changing the model. We may perform the behavior per batch, per epoch or at the start and end of job. `tf.keras.callbacks.Callback` is an abstract base class and has methods to perform the behavior at different call frequency, such as `on_bath_end`, `on_epoch_end` and so on. So, we adopt the interfaces of `tf.keras.callbacks.Callback` to customize the behavior of model in ElasticDL. 
+
+Now we have implemented some modules similar to callback in ElasticDL, such as `LearningRateScheduler` and `PredictionOutputsProcessor`. And users should write definitions in the model definition file for each module like:
+```python
+def custom_model():
+    ...
+
+# Adjust the learning rate according to iteration steps
+def learning_rate_scheduler(model_version):
+    if model_version < 5000:
+        return 0.0003
+    elif model_version < 12000:
+        return 0.0002
+    else:
+        return 0.0001
+
+# Process the prediction outputs for each batch
+class PredictionOutputsProcessor(BasePredictionOutputsProcessor):
+    ...
+```
+There will be multiple definition APIs for users to define different behaviors of model. The interfaces of APIs may be different. It is more convenient for users to define those behaviors using `tf.keras.callbacks.Callback`. 
+
+Some use cases we observed that we want to support are:
+* Case 1: Callback similar to `PredictionOutputsProcessor` that is executed after prediction outputs are made.
+* Case 2: Callback to modulate learning rate as we are currently doing in `LearningRateScheduler`.
+* Case 3: Callback to write additional summaries to TensorBoard after each evaluation job completes.
+* Case 4: Callback to perform early stopping when certain conditions are met. For example, the metrics are met after an evaluation job.
+* Case 5: Callback to export model using SavedModel after the training is completed.
+* Case 6: Callback to upload model to remote storage after the training is completed.
+
+The following, we will design how to define and implement callbacks to support those cases.
```

Comment:
The following, we will design how to define and implement callbacks to support those cases.
=>
Next we will design how to define and implement callbacks to support those cases above.